src/login/loginCoordinator.ts

```diff
@@ -0,0 +1,286 @@
+			storedToken = auth?.token;
+			if (storedToken) {
+				client.setSessionToken(storedToken);
+			}
+		}
+
+		// Attempt authentication with current credentials (token or mTLS)
+		try {
+			const user = await client.getAuthenticatedUser();
+			// Return the token that was used (empty string for mTLS since
+			// the `vscodessh` command currently always requires a token file)
+			return { success: true, token: storedToken ?? "", user };
+		} catch (err) {
+			if (needsToken) {
+				// For token auth: silently continue to prompt for new credentials
+			} else {
+				// For mTLS: show error and abort (no credentials to prompt for)
+				const message = getErrorMessage(err, "no response from the server");
+				if (isAutoLogin) {
+					this.logger.warn("Failed to log in to Coder server:", message);
+				} else {
+					this.vscodeProposed.window.showErrorMessage(
+						"Failed to log in to Coder server",
+						{
+							detail: message,
+							modal: true,
+							useCustom: true,
+						},
+					);
+				}
+				return { success: false };
+			}
+		}
+
+		const result = await this.loginWithToken(client);
+		return result;
+	}
+
+	/**
+	 * Session token authentication flow.
+	 */
+	private async loginWithToken(client: CoderApi): Promise<LoginResult> {
+		const url = client.getAxiosInstance().defaults.baseURL;
+		if (!url) {
+			throw new Error("No base URL set on REST client");
+		}
+		// This prompt is for convenience; do not error if they close it since
+		// they may already have a token or already have the page opened.
+		await vscode.env.openExternal(vscode.Uri.parse(`${url}/cli-auth`));
+
+		// For token auth, start with the existing token in the prompt or the last
+		// used token.  Once submitted, if there is a failure we will keep asking
+		// the user for a new token until they quit.
+		let user: User | undefined;
+		const validatedToken = await vscode.window.showInputBox({
+			title: "Coder API Key",
+			password: true,
+			placeHolder: "Paste your API key.",
+			ignoreFocusOut: true,
+			validateInput: async (value) => {
+				if (!value) {
+					return null;
+				}
+				client.setSessionToken(value);
+				try {
+					user = await client.getAuthenticatedUser();
+				} catch (err) {
+					// For certificate errors show both a notification and add to the
+					// text under the input box, since users sometimes miss the
+					// notification.
+					if (err instanceof CertificateError) {
+						err.showNotification();
+						return {
+							message: err.x509Err || err.message,
+							severity: vscode.InputBoxValidationSeverity.Error,
+						};
+					}
+					// This could be something like the header command erroring or an
+					// invalid session token.
+					const message = getErrorMessage(err, "no response from the server");
+					return {
+						message: "Failed to authenticate: " + message,
+						severity: vscode.InputBoxValidationSeverity.Error,
+					};
+				}
+			},
+		});
+
+		if (user === undefined || validatedToken === undefined) {
```

Comment:
Could these be null?  Not sure about `getAuthenticatedUser()` but I guess in theory the body could be null.  And the token could be an empty string?

Reply:
lol unrelated to your PR since this was already a problem, but I just had an issue where I forgot to make my proxied test deployment public so it was trying to fetch the user and it got html back, then it tried to do `user.roles` but `user` is a string...

Maybe at some point we will want to validate that `user` is an actual user object

Reply:
Yuuup too true, too true.  But, if we validate the user at least (which we always fetch) then we have at least some confidence that the rest will work.